### ISSUE TITLE
memcache protocol is missing set_multi and get_multi

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3171]
|<img alt="kvogt's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| kvogt reported|
|-|-|
|Trac ID|trac#3171|
|Type|enhancement|
|Created|2008-04-07 21:46:16Z|
|Branch|https://github.com/twisted/twisted/tree/memcache-get-multiple-3171|

These functions are a part of the standard python memcache library, and make it much easier to work with multiple keys.

Attachments:

* [memcache_multi_support](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/dd7/dd792a02ad2ffc2e4dfcf49ff075ec1f405ffc4a/9a7c349b04692807e508d55a8cafa3833bd28b72) (7168 bytes) - added by kvogt on 2008-04-07 21:46:56Z - Patch
* [memcache_multi_support.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/dd7/dd792a02ad2ffc2e4dfcf49ff075ec1f405ffc4a/0e76ae0da2fb5e776d91449ffdb558e6d2afc760.diff) (7168 bytes) - added by kvogt on 2008-04-07 21:57:28Z - Patch with .diff extension

<details><summary>Searchable metadata</summary>

```
trac-id__3171 3171
type__enhancement enhancement
reporter__kvogt kvogt
priority__normal normal
milestone__ 
branch__branches_memcache_get_multiple_3171 branches/memcache-get-multiple-3171
branch_author__therve therve
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1207604776000000 1207604776000000
changetime__1248662499000000 1248662499000000
version__None None
owner__ 
cc__therve cc__oubiwann
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve set owner to kvogt|
|-|-|

Thanks a lot for this feature. That can be useful indeed. Some comments:
 * the methods on the object should have a more explicit name (like getMultiple and setMultiple)
 * the docstring of set_multi is wrong (for flags and expireTime arguments)
 * same for the return type of get_multi
 * get_multi should support the withIdentifier keywords
 * get_multi could be used with the *args syntax
 * it needs some tests, for the error cases for example.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve commented|
|-|-|

(In [27168]) Branching to 'memcache-get-multiple-3171'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve removed owner|
|-|-|

This is ready to review in the attached branch. I only implemented getMultiple, as there is no protocol level implementation of multiple set.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set owner to @therve|
|-|-|

Cool.  A few simple things:

  1. Please add `@since` to `getMultiple`.
  1. In the docstring for `getMultiple`, I think _with C{keys} as keys_ would be clearer as _with the elements of C{keys} as keys_.
  1. One line is uncovered, the first raise in `cmd_VALUE`

I'm curious about what happens if there is a key with a space in it.  I guess the protocol forbids that?  Should our memcache api locally reject such values?  That's probably not a new issue, since lots of existing APIs operate on keys, but seeing a list of strings joined together with a space made me think about that case.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_8"></a>|@therve removed owner|
|-|-|

Thanks for the review, I've handled the 3 points.

You're right, keys can't contain spaces (any control characters, really). Other libraries seem to check that every character is above 32 strict (and under 127, although some tests show it may work). We should add those checks, but I'd like to postpone to a proper ticket if you don't mind.

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_9"></a>|@jonathanj set owner to @therve|
|-|-|

These changes looks good. However, I noticed that `MemCacheProtocol.get` and `MemCacheProtocol.getMultiple` look almost identical, it might be nice to refactor the common code out.

Apart from that minor concern, I'd say file a new ticket for the key name restrictions and merge this one.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_10"></a>|@therve set status to `closed`|
|-|-|

(In [27178]) Merge memcache-get-multiple-3171

Author: therve
Reviewers: exarkun, jonathanj
Fixes [#3171](https://github.com/twisted/twisted/issues/3171)

Add a getMultiple command on the MemCacheProtocol, to fetch several keys in one
call.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun set status to `reopened`|
|-|-|

(In [27185]) Revert r27178 - Python 2.3 incompatibility

Reopens [#3171](https://github.com/twisted/twisted/issues/3171)

This changeset introduced a use of the generator expression syntax, not supported
on Python 2.3.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

(In [27186]) Switch from generator expressions to list comprehensions

refs [#3171](https://github.com/twisted/twisted/issues/3171)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_14"></a>|@therve set owner to @exarkun|
|-|-|

Hum, sorry about that and fix for the fix. +1.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun set status to `closed`|
|-|-|

(In [27194]) Merge memcache-get-multiple-3171

Author: therve
Reviewers: exarkun, jonathanj
Fixes: [#3171](https://github.com/twisted/twisted/issues/3171)

Add a `getMultiple` method to `MemCacheProtocol`, to fetch several keys in one call.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
